### DISCUSSION TITLE
Get Other Claims From Token

### DISCUSSION THREAD

**Discussion by jrunyan77** [paolo-rossi/delphi-jose-jwt#81] [Q&A] [answered]
I was able to create a token

```
  LKey := TJWK.Create(FPublicPEM);

  // Unpack and verify the token!
  LToken := TJOSE.Verify(LKey, TokenMmo.Lines.Text);`
```
Then I use a Consumer for verification.
```  Consumer :=
    TJOSEConsumerBuilder
      .NewConsumer
      .SetClaimsClass(TJWTClaims)
      .SetVerificationKey(FPublicPEM)
      .SetExpectedAudience(True, ['urn:gov:dot:openidconnect.profiles:sp:sso:fmcsa:test_oauth'])
      .SetRequireIssuedAt
      .SetEvaluationTime(IncDay(Now, -1))
      .Build;
```
But now I want to check a claim called "nonce". But I could not find a way to set it in the consumer.

So then I tried to get it from the Token.

1st 
`LToken.ClaimsAs<TJOSEBase>.JSON.Values['nonce'];`

2nd
`(LToken.Claims as TJoseBase).JSON.As<TJOSEBase>.JSON.Values['nonce'];`

Last, I tried to get the entire JSON string.
`(LToken.Claims as TJoseBase).JSON.ToString;`

Do I have to create a descendant of TJWTClaims and add nonce as a property?

**Comment by jrunyan77** [comment]
@paolo-rossi,

I found that I could create a custom claims class and add and access new claims there. But there does not seem to be a way to extend the Consumer class to add claims and validators.

Is there a reason that the functions to do this are marked private?

Thanks,

John

**Comment by paolo-rossi** [comment, marked as answer]
Hello @jrunyan77 

the validation of custom claims can be done through a consumer adding custom validators, I've extended the CustomClaims sample to show how.

hope this will address the issue..

Paolo.
